File: beecrowd/2962/solution.cpp
#include <cstdio>
#include <cmath>


class UnionFind {
    int parent[1004];
    int rank[1004];
public:
    UnionFind(int n) {
        for (int i = 0; i < n; i++){
            parent[i] = i;
            rank[i] = 1;
        }
    }

    int find(int n) {
        if (parent[n] != n) {
            parent[n] = find(parent[n]);
        }
        return parent[n];
    }

    void merge(int a, int b) {
        int parent_a = find(a);
        int parent_b = find(b);

        if (rank[parent_a] > rank[parent_b]) {
            parent[parent_b] = parent_a;
        } else if (rank[parent_a] < rank[parent_b]) {
            parent[parent_a] = parent_b; 
        } else {
            parent[parent_b] = parent_a;
            rank[parent_a]++;
        }
    }

    bool connected(int a, int b) {
        return find(a) == find(b);
    }
};


double dist(int p1[3], int p2[3]) {
    return sqrt(
        (p1[0] - p2[0]) * (p1[0] - p2[0]) +
        (p1[1] - p2[1]) * (p1[1] - p2[1])
    );
}


int main (int argc, char *argv[]) {
    int m, n, k;
    int sensors[1000][3];

    scanf("%d %d %d", &m, &n, &k);
    /* Nós 0, 1, 2, e 3 são reservados para o topo, a direita, em baixo, e a
     * esquerda respectivamente. */
    auto union_find = UnionFind(k + 4);

    for (int i = 0; i < k; i++){
        scanf("%d %d %d", &sensors[i][0], &sensors[i][1], &sensors[i][2]);

        // Sensor encherga a parede do topo
        if (sensors[i][1] + sensors[i][2] >= n)
            union_find.merge(0, i + 4);

        // Sensor encherga a parede da direita
        if (sensors[i][0] + sensors[i][2] >= m)
            union_find.merge(1, i + 4);

        // Sensor encherga a parede de baixo
        if (sensors[i][1] - sensors[i][2] <= 0)
            union_find.merge(2, i + 4);

        // Sensor encherga a parede da esquerda
        if (sensors[i][0] - sensors[i][2] <= 0)
            union_find.merge(3, i + 4);
    }

    for (int i = 0; i < k; i++) {
        for (int j = i + 1; j < k; j++){
            if (dist(sensors[i], sensors[j]) <= sensors[i][2] + sensors[j][2])
                union_find.merge(i + 4, j + 4);
        }
    }

    if (union_find.connected(0, 1) || union_find.connected(0, 2) ||
        union_find.connected(3, 1) || union_find.connected(3, 2))
        printf("N\n");
    else
        printf("S\n");

    return 0;
}

File: beecrowd/2962/solution.py
from sys import stdin
from math import dist

class UnionFind:
    def __init__(self, n: int) -> None:
        self.rank = [0] * n
        self.parent = [i for i in range(n)]

    def find(self, n: int) -> int:
        # Comprime o caminho para o elemento pai
        if n != self.parent[n]:
            self.parent[n] = self.find(self.parent[n])
        return self.parent[n]

    def union(self, a: int, b: int) -> None:
        root_a = self.find(a)
        root_b = self.find(b)

        if self.rank[root_a] > self.rank[root_b]:
            self.parent[root_b] = root_a

        elif self.rank[root_b] > self.rank[root_a]:
            self.parent[root_a] = root_b

        else:
            self.parent[root_b] = root_a
            self.rank[root_a] += 1

    def connected(self, a: int, b: int) -> bool:
        return self.find(a) == self.find(b)



def main():
    m, n, k = map(int, stdin.readline().split())
    # Nós 0, 1, 2, e 3 são reservados respectivamente para o topo, a direita
    # em baixo, e a esquerda.
    union_find = UnionFind(k + 4)

    sensors = [None] * k
    for i, line in enumerate(stdin.readlines()):
        x, y, s = map(int, line.split())
        sensors[i] = (x, y, s) # type: ignore

        # Sensor enxerga a parede do topo
        if y + s >= n:
            union_find.union(0, i + 4)
        # Sensor encerga a parede da direita
        if x + s >= m:
            union_find.union(1, i + 4)
        # Sensor enxerga a parede de baixo
        if y - s <= 0:
            union_find.union(2, i + 4)
        # Sensor enxerga a parede da esquerda
        if x - s <= 0:
            union_find.union(3, i + 4)


    for i, (x1, y1, s1) in enumerate(sensors): # type: ignore
        for j, (x2, y2, s2) in enumerate(sensors[i+1:]): # type: ignore
            if dist((x1, y1), (x2, y2)) <= s1 + s2:
                idx_i = i + 4
                idx_j = i + j + 5

                union_find.union(idx_i, idx_j)


    if union_find.connected(0, 1) or union_find.connected(0, 2) or \
       union_find.connected(3, 1) or union_find.connected(3, 2):
        print("N")
    else:
        print("S")
 


if __name__ == "__main__":
    main()
